### PR TITLE
Fixed display of optional and rest parameters created as result of combining union parameters

### PR DESCRIPTION
**Pull request by Andarist**
fixes https://github.com/microsoft/TypeScript/issues/55574

### CODE REVIEW

File: tests/cases/fourslash/completionEntryForUnionMethod.ts [hunk 1]
```diff
@@ -3,7 +3,7 @@
 ////var y: Array<string>|Array<number>;
 ////y.map/**/(
 
-const text = "(method) Array<T>.map<unknown>(callbackfn: ((value: string, index: number, array: string[]) => unknown) & ((value: number, index: number, array: number[]) => unknown), thisArg: any): unknown[]";
+const text = "(method) Array<T>.map<unknown>(callbackfn: ((value: string, index: number, array: string[]) => unknown) & ((value: number, index: number, array: number[]) => unknown), thisArg?: any): unknown[]";
```
Comment (Andarist) [inline tests/cases/fourslash/completionEntryForUnionMethod.ts:6]: From what I can tell - this is the correct change. It matches the display of the same method accessed on a single array (this one accesses it on a union of arrays):
```ts
var y: Array<string>
y.map(
// (method) Array<string>.map<unknown>(callbackfn: (value: string, index: number, array: string[]) => unknown, thisArg?: any): unknown[]
```

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -13201,12 +13201,13 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
...
                 SymbolFlags.FunctionScopedVariable | (isOptional && !isRestParam ? SymbolFlags.Optional : 0),
                 paramName || `arg${i}` as __String,
+                isRestParam ? CheckFlags.RestParameter : isOptional ? CheckFlags.OptionalParameter : 0,
             );
             paramSymbol.links.type = isRestParam ? createArrayType(unionParamType) : unionParamType;
             params[i] = paramSymbol;
         }
         if (needsExtraRestElement) {
-            const restParamSymbol = createSymbol(SymbolFlags.FunctionScopedVariable, "args" as __String);
+            const restParamSymbol = createSymbol(SymbolFlags.FunctionScopedVariable, "args" as __String, CheckFlags.RestParameter);
```
Comment (Andarist) [inline src/compiler/checker.ts:13210]: to give context: those are later checked by `symbolToParameterDeclaration` here: https://github.dev/microsoft/TypeScript/blob/5d04803196899fab535103a5114bc9bf8d04dbd9/src/compiler/checker.ts#L7615-L7618

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -13201,12 +13201,13 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             const paramSymbol = createSymbol(
                 SymbolFlags.FunctionScopedVariable | (isOptional && !isRestParam ? SymbolFlags.Optional : 0),
                 paramName || `arg${i}` as __String,
+                isRestParam ? CheckFlags.RestParameter : isOptional ? CheckFlags.OptionalParameter : 0,
```
Comment (jakebailey) [inline src/compiler/checker.ts:13204]: `0` should probably be `undefined` or `CheckFlags.None`.

**Comment by jakebailey** [comment]
@typescript-bot test top100
@typescript-bot user test this
@typescript-bot user test tsserver
@typescript-bot test tsserver top100
@typescript-bot run dt
@typescript-bot perf test this

File: src/compiler/checker.ts (hunk 3 above)
Comment (Andarist) [reply src/compiler/checker.ts:13204]: I followed the existing "convention" here with that `0` :p the first argument here is passed in this very way

Comment (jakebailey) [reply src/compiler/checker.ts:13204]: Er, well, that's `SymbolFlags`, but yeah that also doesn't use `None` for some reason, boo

Comment (Andarist) [reply src/compiler/checker.ts:13204]: I can adjust both places if you'd prefer that

Comment (jakebailey) [reply src/compiler/checker.ts:13204]: Eh, it's fine; commenting out `.None` shows 5 other places we used zero and none where we used `None`.

**Review by jakebailey — APPROVED** [review]
This seems correct to me.

**Comment by sandersn** [comment]
@jakebailey is this ready to merge now?

**Comment by jakebailey** [comment]
I'd say so but I typically don't merge PRs that aren't assigned to me just in case 😅

**Review by weswigham — APPROVED** [review]

**Pull request merged**